x/distribution/keeper/allocation.go

```diff
@@ -40,32 +40,56 @@ func (k Keeper) AllocateTokens(ctx context.Context, totalPreviousPower int64, bo
 		return k.FeePool.Set(ctx, feePool)
 	}
 
-	// calculate fraction allocated to validators
-	remaining := feesCollected
+	// Get community tax and Nakamoto bonus ratio η
 	communityTax, err := k.GetCommunityTax(ctx)
 	if err != nil {
 		return err
 	}
+	params, err := k.Params.Get(ctx)
+	if err != nil {
+		return err
+	}
+	nakamotoCoefficient := params.NakamotoBonusCoefficient // the nakamoto bonus coefficient (e.g., 0.05 means 5% NB, 95% PR)
 
+	// Compute total validator rewards (after community tax)
 	voteMultiplier := math.LegacyOneDec().Sub(communityTax)
-	feeMultiplier := feesCollected.MulDecTruncate(voteMultiplier)
+	validatorTotalReward := feesCollected.MulDecTruncate(voteMultiplier)
 
-	// allocate tokens proportionally to voting power
-	//
-	// TODO: Consider parallelizing later
-	//
-	// Ref: https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r246276376
+	// Split reward into Proportional (PR_i) and Nakamoto Bonus (NB_i)
+	nakamotoBonus := sdk.DecCoins{}
+	if params.NakamotoBonusEnabled {
+		nakamotoBonus = validatorTotalReward.MulDecTruncate(nakamotoCoefficient)
+	}
+	proportionalReward := validatorTotalReward.Sub(nakamotoBonus)
+
+	bondDenom, err := k.stakingKeeper.BondDenom(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Compute per-validator fixed Nakamoto bonus
+	numValidators := int64(len(bondedVotes))
+	nbPerValidator := sdk.NewDecCoinFromDec(bondDenom, math.LegacyZeroDec())
+	if numValidators > 0 && len(nakamotoBonus) > 0 {
```

Comment:
```suggestion
	if numValidators > 0 && !nakamotoBonus.IsZero() {
```

This avoids some operations if the nakamotoCoefficient is 0.

Reply:
@Pantani Please comment why you marked this comment as resolved.

Reply:
This logic does not exist in the code anymore. Can you check again, please?!

Reply:
wrong statement from me, the comment was resolved properly.

Reply:
thanks for the review